x-pack/plugins/viz_editor/server/query/select_operations.ts

```diff
@@ -0,0 +1,232 @@
+          },
+        },
+      };
+    },
+  },
+
+  count: {
+    getName(op: AvgOperation) {
+      return op.alias || 'count';
+    },
+    toEsAgg(op: CountOperation) {
+      // We perform a count in this way so that it
+      // is consistently represented regardless of
+      // other query concerns (groupby, histogram, etc)
+      return {
+        [op.alias!]: {
+          value_count: {
+            field: '_id',
+          },
+        },
+      };
+    },
+  },
+
+  avg: defineBasicAgg('avg'),
+  sum: defineBasicAgg('sum'),
+};
+
+/**
+ * Basic aggregations seem to pretty much share the same shape,
+ * so this is a little helper to generate the definition for
+ * the selectOperations map.
+ */
+function defineBasicAgg(aggName: string) {
+  return {
+    getName(op: { alias?: string; argument: string }) {
+      return op.alias || `${aggName}_${op.argument}`;
+    },
+    toEsAgg(op: { alias: string; argument: string }) {
+      return {
+        [op.alias]: {
+          [aggName]: {
+            field: op.argument,
+          },
+        },
+      };
+    },
+  };
+}
+
+/**
+ * Determine if the specified operation is an aggregation
+ */
+function isAggregationOperation(select: SelectOperation): boolean {
+  const definition = selectOperations[select.operation];
+  return !!definition && !!definition.toEsAgg;
+}
+
+/**
+ * Determine if the specified query has aggregation selections
+ */
+export function hasAggregations(query: Query): boolean {
+  return query.select.some(isAggregationOperation);
+}
+
+/**
+ * Convert the specified operation to an Elasticsearch aggregation
+ */
+function toEsAgg(op: SelectOperation) {
+  const { operation } = op;
+  const definition = selectOperations[operation];
+
+  if (!definition || !definition.toEsAgg) {
+    throw new Error(`Unrecognized aggregation operation: ${operation}`);
+  }
+
+  return definition.toEsAgg(op);
+}
+
+/**
+ * Convert the specified select operations to a nested groupby / histogram tree
+ */
+function toNestedQuery(query: Query, select: SelectOperation[], leafAggregations: any) {
+  // Recursively build a nested groupby / histogram tree
+  function nestGroupings([chunk, ...chunks]: Array<ChunkResult<SelectOperation>>): any {
+    const operations = chunk && chunk.values;
+    if (!operations || !operations.length) {
+      return leafAggregations || {};
+    }
+
+    const [op] = operations;
+    const operationName = op.operation;
+    const operation = selectOperations[operationName];
+
+    if (!operation || !operation.toNestedQuery) {
+      throw new Error(`Cannot nest by ${operationName}`);
+    }
+
+    return operation.toNestedQuery(query, operations, () => nestGroupings(chunks));
+  }
+
+  return nestGroupings(chunkBy(c => c.operation, select));
+}
+
+/**
+ * Convert the specified query into an Elasticsearch query that has aggregations
+ */
+function buildAggregations(query: Query, esQuery: any) {
+  const [aggs, cols] = partition(isAggregationOperation, query.select);
```

Comment:
If it's this important to split the fields in `query.select` into just "aggs" vs "columns" maybe these should be split up at the query level as well

Reply:
That may be true. In the view model, we probably don't want it split, as the user can fairly fluidly change fields: "I want this to be bytes... no, make that sum of bytes, nah let's make it an average..." so, for the view model, we want to have a flat set of fields. This is mostly a reflection of that. That, and if we ever do use this as a foundation for queries in other parts of Kibana, I think it's easier from an external consumption / API layer to just say, "I want airport, sum(ticket_price)" and not have to move things around structurally to make the change from "ticket_price" to "sum(ticket_price)"